File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/test/LocalizationTest.java
package org.firstinspires.ftc.teamcode.test;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.dashboard.telemetry.MultipleTelemetry;
import com.acmerobotics.roadrunner.geometry.Pose2d;
import com.qualcomm.robotcore.eventloop.opmode.Disabled;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;

import org.firstinspires.ftc.teamcode.hardware.drive.AcmeHolonomicDrive;

@Config
@TeleOp(group = "Test")
@Disabled
public class LocalizationTest extends LinearOpMode {

    public static double VX_WEIGHT = 1;
    public static double VY_WEIGHT = 1;
    public static double OMEGA_WEIGHT = 1;

    @Override
    public void runOpMode() throws InterruptedException {

        telemetry = new MultipleTelemetry(telemetry, FtcDashboard.getInstance().getTelemetry());

        AcmeHolonomicDrive drive = new AcmeHolonomicDrive(hardwareMap);

        waitForStart();

        while (!isStopRequested()) {

            Pose2d baseVel = new Pose2d(
                    -gamepad1.left_stick_y,
                    -gamepad1.left_stick_x,
                    -gamepad1.right_stick_x
            );

            Pose2d vel;
            if (Math.abs(baseVel.getX()) + Math.abs(baseVel.getY()) + Math.abs(baseVel.getHeading()) > 1) {

                // re-normalize the powers according to the weights
                double denom = VX_WEIGHT * Math.abs(baseVel.getX())
                        + VY_WEIGHT * Math.abs(baseVel.getY())
                        + OMEGA_WEIGHT * Math.abs(baseVel.getHeading());
                vel = new Pose2d(
                        VX_WEIGHT * baseVel.getX(),
                        VY_WEIGHT * baseVel.getY(),
                        OMEGA_WEIGHT * baseVel.getHeading()
                ).div(denom);

            } else
                vel = baseVel;

            drive.setDrivePower(vel);

            drive.update();

            Pose2d poseEstimate = drive.getPoseEstimate();
            telemetry.addData("x", poseEstimate.getX());
            telemetry.addData("y", poseEstimate.getY());
            telemetry.addData("heading", poseEstimate.getHeading());
            telemetry.update();

        }

    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/drive/opmode/LocalizationTest.kt
package org.firstinspires.ftc.teamcode.drive.opmode;

import com.acmerobotics.roadrunner.drive.Drive
import com.arcrobotics.ftclib.gamepad.GamepadKeys
import com.outoftheboxrobotics.photoncore.Photon;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.scrapmetal.quackerama.control.Pose2d
import com.scrapmetal.quackerama.control.Rotation2d

import org.firstinspires.ftc.teamcode.drive.SampleMecanumDrive;
import org.firstinspires.ftc.teamcode.seventh.opmode.teleop.Gigapad
import org.firstinspires.ftc.teamcode.seventh.robot.hardware.Globals;
import org.firstinspires.ftc.teamcode.seventh.robot.hardware.Robot
import org.firstinspires.ftc.teamcode.seventh.robot.subsystem.DriveSubsystem
import org.firstinspires.ftc.teamcode.seventh.robot.vision.Vision;
import java.lang.Math.toDegrees
import kotlin.math.pow
import kotlin.math.sign

/**
 * This is a simple teleop routine for testing localization. Drive the robot around like a normal
 * teleop routine and make sure the robot's estimated pose matches the robot's actual pose (slight
 * errors are not out of the ordinary, especially with sudden drive motions). The goal of this
 * exercise is to ascertain whether the localizer has been configured properly (note: the pure
 * encoder localizer heading may be significantly off if the track width has not been tuned).
 */
@Photon
@TeleOp(group = "drive")
public class LocalizationTest : LinearOpMode() {
    override fun runOpMode() {
        Globals.alliance = Globals.Alliance.RED
        val gigapad = Gigapad(gamepad1)
        Robot.init(hardwareMap, telemetry, gigapad, null)
        val drive = DriveSubsystem(hardwareMap);
        val vision = Vision(hardwareMap);

        vision.initAtag()

        drive.startIMUThread(this);
        waitForStart();
        drive.setPoseEstimate(Pose2d())

        while (!isStopRequested) {
            Robot.read(drive)
            drive.updateOpen(
                Pose2d(
                    gigapad.leftY,
                    -gigapad.leftX,
                    -gigapad.rightX.let { it.sign*it.pow(2) },
                ),
                fieldOriented = false,
                headingPID = false,
            )

            if (gigapad.wasJustPressed(GamepadKeys.Button.A)) {
                vision.getPosition(drive.getPoseEstimate().heading)?.let {
                    drive.setPoseEstimate(Pose2d(it, drive.getPoseEstimate().heading))
                }
            }

            telemetry.addData("hz", 1 / Robot.getDt());
            with (drive.getPoseEstimate()) {
                telemetry.addData("x", position.x);
                telemetry.addData("y", position.y);
                telemetry.addData("heading", toDegrees(heading.theta))
            }
            Robot.write(drive)
        }
    }
}
